File: src/kitchen_sync/adapters/docker.py
"""Docker container environment adapter."""

from __future__ import annotations

import subprocess
from pathlib import Path

from kitchen_sync.adapters.base import Environment


class DockerEnvironment(Environment):
    """Adapter that deploys via docker exec/cp to a running container."""

    def __init__(self, image: str, targets: dict | None = None):
        super().__init__(targets)
        self.image = image
        self._container_id: str | None = None

    @property
    def container_id(self) -> str | None:
        """Find the running container ID for this image."""
        if self._container_id is None:
            result = subprocess.run(
                [
                    "docker",
                    "ps",
                    "--filter",
                    f"ancestor={self.image}",
                    "--format",
                    "{{.ID}}",
                ],
                capture_output=True,
                text=True,
            )
            ids = result.stdout.strip().splitlines()
            if ids:
                self._container_id = ids[0]
        return self._container_id

    def is_available(self) -> bool:
        return self.container_id is not None

    def run(self, cmd: list[str], **kwargs) -> subprocess.CompletedProcess:
        cid = self.container_id
        if not cid:
            raise RuntimeError(f"No running container for image {self.image}")
        return subprocess.run(
            ["docker", "exec", cid] + cmd,
            capture_output=True,
            text=True,
            **kwargs,
        )

    def read_file(self, path: str) -> str | None:
        result = self.run(["cat", path])
        if result.returncode == 0:
            return result.stdout
        return None

    def list_files(self, path: str) -> list[str]:
        result = self.run(["find", path, "-type", "f"])
        if result.returncode != 0:
            return []
        files = []
        for line in result.stdout.strip().splitlines():
            if line.startswith(path):
                rel = line[len(path) :].lstrip("/")
                if rel:
                    files.append(rel)
        return sorted(files)

    def deploy(
        self,
        staging_dir: Path,
        target_dir: str,
        sync_paths: list[str],
    ) -> list[str]:
        cid = self.container_id
        if not cid:
            raise RuntimeError(f"No running container for image {self.image}")

        # Ensure target dir exists
        self.run(["mkdir", "-p", target_dir])

        deployed = []
        for sp in sync_paths:
            src = staging_dir / sp
            if not src.exists():
                continue

            remote_path = f"{target_dir}/{sp}"

            # docker cp local_path container:remote_path
            subprocess.run(
                ["docker", "cp", str(src), f"{cid}:{remote_path}"],
                check=True,
                capture_output=True,
            )
            deployed.append(sp)

        return deployed

    def clean(self, target_dir: str, sync_paths: list[str]) -> None:
        for sp in sync_paths:
            self.run(["rm", "-rf", f"{target_dir}/{sp}"])

    @property
    def display_name(self) -> str:
        short_id = self.container_id[:12] if self.container_id else "?"
        return f"docker container {short_id}"

File: src/adapters/docker.ts
import { spawnSync } from "child_process";
import { Environment, type RunResult } from "./base.js";

const DOCKER_IMAGE_RE = /^[a-zA-Z0-9][a-zA-Z0-9._\/:@-]*$/;

export class DockerEnvironment extends Environment {
  private image: string;
  private _containerId: string | null = null;

  constructor(image: string) {
    super();
    if (!DOCKER_IMAGE_RE.test(image) || /\0/.test(image)) {
      throw new Error(`Invalid Docker image name: ${image}`);
    }
    this.image = image;
  }

  get containerId(): string | null {
    if (this._containerId === null) {
      const result = spawnSync(
        "docker",
        ["ps", "--filter", `ancestor=${this.image}`, "--format", "{{.ID}}"],
        { encoding: "utf-8" },
      );
      const ids = (result.stdout ?? "").trim().split("\n").filter(Boolean);
      if (ids.length > 0) {
        this._containerId = ids[0];
      }
    }
    return this._containerId;
  }

  // Exposed for testing
  set containerId(id: string | null) {
    this._containerId = id;
  }

  isAvailable(): boolean {
    return this.containerId !== null;
  }

  run(cmd: string[]): RunResult {
    const cid = this.containerId;
    if (!cid) {
      throw new Error(`No running container for image ${this.image}`);
    }
    const result = spawnSync("docker", ["exec", cid, ...cmd], { encoding: "utf-8" });
    return {
      stdout: result.stdout ?? "",
      stderr: result.stderr ?? "",
      exitCode: result.status ?? 1,
    };
  }

  readFile(filePath: string): string | null {
    const result = this.run(["cat", filePath]);
    if (result.exitCode === 0) {
      return result.stdout;
    }
    return null;
  }

  listFiles(dirPath: string): string[] {
    const result = this.run(["find", dirPath, "-type", "f"]);
    if (result.exitCode !== 0) return [];
    const files: string[] = [];
    for (const line of result.stdout.trim().split("\n")) {
      if (line.startsWith(dirPath)) {
        const rel = line.slice(dirPath.length).replace(/^\//, "");
        if (rel) files.push(rel);
      }
    }
    return files.sort();
  }

  deploy(stagingDir: string, targetDir: string, syncPaths: string[]): string[] {
    const cid = this.containerId;
    if (!cid) {
      throw new Error(`No running container for image ${this.image}`);
    }

    // Ensure target dir exists
    this.run(["mkdir", "-p", targetDir]);

    const deployed: string[] = [];
    for (const sp of syncPaths) {
      const src = `${stagingDir}/${sp}`;
      // Check if source exists using node fs (staging is local)
      try {
        const { statSync } = require("fs");
        statSync(src);
      } catch {
        continue;
      }

      const remotePath = `${targetDir}/${sp}`;
      spawnSync("docker", ["cp", src, `${cid}:${remotePath}`], { encoding: "utf-8" });
      deployed.push(sp);
    }

    return deployed;
  }

  clean(targetDir: string, syncPaths: string[]): void {
    for (const sp of syncPaths) {
      this.run(["rm", "-rf", `${targetDir}/${sp}`]);
    }
  }

  get displayName(): string {
    const shortId = this._containerId ? this._containerId.slice(0, 12) : "?";
    return `docker container ${shortId}`;
  }
}
